native/dll/Win32DLL/WindowsServiceControl.cpp

```diff
@@ -651,47 +647,37 @@ GetAccountSid(
     // 
     // initial memory allocations
     // 
-    if((*Sid=HeapAlloc(
-                    GetProcessHeap(),
-                    0,
-                    cbSid
-                    )) == NULL) __leave;
-
-    if((ReferencedDomain=(LPTSTR)HeapAlloc(
-                    GetProcessHeap(),
-                    0,
-                    cchReferencedDomain * sizeof(TCHAR)
-                    )) == NULL) __leave;
+    if((*Sid=HeapAlloc(GetProcessHeap(), 0, cbSid)) == nullptr) __leave;
+
+    if((ReferencedDomain=(LPTSTR)HeapAlloc(GetProcessHeap(), 0, cchReferencedDomain * sizeof(TCHAR))) == NULL) __leave;
 
     // 
     // Obtain the SID of the specified account on the specified system.
     // 
-    while(!LookupAccountName(
-                    SystemName,         // machine to lookup account on
-                    AccountName,        // account to lookup
-                    *Sid,               // SID of interest
-                    &cbSid,             // size of SID
-                    ReferencedDomain,   // domain account was found on
-                    &cchReferencedDomain,
-                    &peUse
-                    )) {
-        if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
-            // 
-            // reallocate memory
-            // 
-            if((*Sid=HeapReAlloc(
-                        GetProcessHeap(),
-                        0,
-                        *Sid,
-                        cbSid
-                        )) == NULL) __leave;
+	if ((Sid != nullptr) && (ReferencedDomain != NULL))
+		while (!LookupAccountName(
+		SystemName,         // machine to lookup account on
+		AccountName,        // account to lookup
+		*Sid,               // SID of interest
+		&cbSid,             // size of SID
+		ReferencedDomain,   // domain account was found on
+		&cchReferencedDomain,
+		&peUse
+	))
+	{
+		if (GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+		{
+			// 
+			// reallocate memory
+			//
+
+#pragma warning(disable:6308) // Static Code Analysis" 'realloc' might return null pointer, will cause the original memory block to be leaked..
```

Comment:
This sounds like something that should be fixed rather than hidden with a pragma warning

Reply:
From what I could tell it's an issue with the Static Code Analysis.
It is checking for a NULL return and we already check that the arguments passed to it are not NULL

Reply:
I looked into this, and what it's referring to is that if the HeapReAlloc function fails, then it will return NULL. In that case, the Sid block we allocated before will still have been allocated...BUT since we set the return value to that, we have then lost that value and cannot deallocate it. So this is a legitimate warning.

Reply:
I had a few goes at this before, without success, but I think I have  it now.

Reply:
Still battling `Using uninitialized memory 'ReferencedDomain'` on this

Reply:
What line does it flag that error on?

For the memory leak, consider changing the HeapRealloc to use a new variable which you copy into the original only upon success.  Ditto for ReferencedDomain.  You probably don’t need the “else” I added there if ”__leave” returns.  I didn’t find its definition yet.

PSID newSid;
if ((newSid = HeapReAlloc(GetProcessHeap(), 0, *Sid, cbSid)) == nullptr)
  __leave;
else
  *Sid = newSid;

Reply:
I'd got to here`			//

			PSID *tmpSid = new PSID;
			*tmpSid = *Sid;
			if ((*tmpSid = HeapReAlloc(GetProcessHeap(), 0, *Sid, cbSid)) == nullptr)
			{
				HeapFree(GetProcessHeap(), 0, *tmpSid);
				*tmpSid = NULL;
				__leave;
			}
			else
			{
				*Sid = tmpSid;
				HeapFree(GetProcessHeap(), 0, *tmpSid);
				*tmpSid = NULL;
			}

			LPTSTR tmpReferencedDomain = nullptr;
			if ((tmpReferencedDomain = (LPTSTR)HeapReAlloc(GetProcessHeap(), 0, ReferencedDomain, cchReferencedDomain * sizeof(TCHAR))) == NULL)
			{
				HeapFree(GetProcessHeap(), 0, tmpReferencedDomain);
				__leave;
			}
			else
			{
				strcpy_s(ReferencedDomain, strlen(ReferencedDomain), tmpReferencedDomain);
				HeapFree(GetProcessHeap(), 0, tmpReferencedDomain);
			}`
and I get the "using Uninitialized memory" warning on `strcpy_s(ReferencedDomain, strlen(ReferencedDomain), tmpReferencedDomain);` if that is `ReferencedDomain = tmpReferencedDomain;` then I get the "using Uninitialized memory" warning down @ 
`    // 
    // Cleanup and indicate failure, if appropriate.
    // 

    HeapFree(GetProcessHeap(), 0, ReferencedDomain);
`

Reply:
A few comments on the above inlined below...

> I'd got to here` //
> 
> ```
> 		PSID *tmpSid = new PSID;
Should instead be 'PSID tmpSid = NULL;', you shouldn't be allocating anything here because this is just temporary storage for a return value. It's also already a pointer (that's what the P prefix means) so you don't need to make it a double pointer (we need to have a double pointer passed into the function because we are using that to return the allocated structure).
> 		*tmpSid = *Sid;
If you did this, and had 'new PSID' above, you would be leaking that allocated PSID object. You don't need to do this anyways because tmpSid is getting assigned on the next line, so any assignment here won't matter at all, just delete this line.
> 		if ((*tmpSid = HeapReAlloc(GetProcessHeap(), 0, *Sid, cbSid)) == nullptr)
Get rid of the asterisk on 'tmpSid' here.
> 		{
> 			HeapFree(GetProcessHeap(), 0, *tmpSid);
This is trying to free the wrong thing. tmpSid will be null, because that's what the conditional checks above. Instead, here you should be freeing '*Sid' and then setting it to NULL on the next line. 
> 			*tmpSid = NULL;
> 			__leave;
> 		}
> 		else
> 		{
> 			*Sid = tmpSid;
> 			HeapFree(GetProcessHeap(), 0, *tmpSid);
Get rid of this line because otherwise you are freeing what you just re-allocated because you set them equal on the above line.
> 			*tmpSid = NULL;
This line is not necessary, but it doesn't hurt either.
> 		}
> 
> 		LPTSTR tmpReferencedDomain = nullptr;
> 		if ((tmpReferencedDomain = (LPTSTR)HeapReAlloc(GetProcessHeap(), 0, ReferencedDomain, cchReferencedDomain * sizeof(TCHAR))) == NULL)
> 		{
> 			HeapFree(GetProcessHeap(), 0, tmpReferencedDomain);
You should be freeing ReferencedDomain here instead since that's the initially allocated thing you'd be leaking if the realloc fails.
> 			__leave;
> 		}
> 		else
> 		{
> 			strcpy_s(ReferencedDomain, strlen(ReferencedDomain), tmpReferencedDomain);
You shouldn't be trying to copy a string here. You should just be putting the allocated memory on the right pointer. So instead this should just be: 'ReferencedDomain = tmpReferencedDomain;'
> 			HeapFree(GetProcessHeap(), 0, tmpReferencedDomain);
Get rid of this line, otherwise you are freeing the memory you just allocated.
> 		}`
> ```
> 
> and I get the "using Uninitialized memory" warning on `strcpy_s(ReferencedDomain, strlen(ReferencedDomain), tmpReferencedDomain);` if that is `ReferencedDomain = tmpReferencedDomain;` then I get the "using Uninitialized memory" warning down @
> ` //
> // Cleanup and indicate failure, if appropriate.
> //
> 
> ```
> HeapFree(GetProcessHeap(), 0, ReferencedDomain);
Put an 'if (ReferencedDomain)' (i.e. not null, that's the shortcut way of doing it) above this and I think that'll clean up the error. It's possible ReferencedDomain is never allocated if the original allocation of Sid fails and that's likely what it is warning you on.
> ```
> 
> `

Reply:
Is this a record?  Looks like I've succeeded in getting pretty much every line wrong :)

Reply:
> It's possible ReferencedDomain is never allocated if the original allocation of Sid fails and that's likely what it is warning you on.

Yes, I can see that this is what's happening, so I assume we'll need to initialize ReferencedDomain up here somehow
 
`BOOL GetAccountSid(LPTSTR SystemName, LPTSTR AccountName, PSID *Sid)

{
    LPTSTR ReferencedDomain = nullptr;
    DWORD cbSid=128;    // initial allocation attempt
`

Reply:
I don't see how we ever end up at the HeapRealloc if we leave the __try block when HeapAlloc fails

Reply:
You shouldn't hit HeapRealloc if the HeapAlloc fails and you _leave the try block. What error are you seeing now and what does the code look like?

Reply:
If I put an 'if (ReferencedDomain), I get a warning as ReferencedDomain can be uninitialized if the HeapAlloc fails.

Doesn't the HeapAlloc have to fail in order to get ERROR_INSUFFICIENT_BUFFER returned from GetLastError(), but we'll never get that becuase we leave the __try?

Reply:
Can you post the code you have for that whole function please?

Reply:
Here you go

```
BOOL GetAccountSid(LPTSTR SystemName, LPTSTR AccountName, PSID *Sid)
{
    LPTSTR ReferencedDomain = NULL;
    DWORD cbSid=128;    // initial allocation attempt
    DWORD cchReferencedDomain=16; // initial allocation size
    SID_NAME_USE peUse;
    BOOL bSuccess=FALSE; // assume this function will fail

   __try {

    // 
    // initial memory allocations
    // 
    if((*Sid=HeapAlloc(GetProcessHeap(), 0, cbSid)) == NULL) __leave;

    if((ReferencedDomain=(LPTSTR)HeapAlloc(GetProcessHeap(), 0, cchReferencedDomain * sizeof(TCHAR))) == NULL) __leave;

    // 
    // Obtain the SID of the specified account on the specified system.
    // 
	while (!LookupAccountName(
	SystemName,         // machine to lookup account on
	AccountName,        // account to lookup
	*Sid,               // SID of interest
	&cbSid,             // size of SID
	ReferencedDomain,   // domain account was found on
	&cchReferencedDomain,
	&peUse
	))
	{
		if (GetLastError() == ERROR_INSUFFICIENT_BUFFER)
		{
			// 
			// reallocate memory
			//

			PSID tmpSid = NULL;
			if ((tmpSid = HeapReAlloc(GetProcessHeap(), 0, *Sid, cbSid)) == NULL)
			{
				HeapFree(GetProcessHeap(), 0, *Sid);
				*Sid = NULL;
				__leave;
			}
			else
			{
				*Sid = tmpSid;
			}

			LPTSTR tmpReferencedDomain = NULL;
			if ((tmpReferencedDomain = (LPTSTR)HeapReAlloc(GetProcessHeap(), 0, ReferencedDomain, cchReferencedDomain * sizeof(TCHAR))) == NULL)
			{
				HeapFree(GetProcessHeap(), 0, ReferencedDomain);
				__leave;
			}
			else
			{
				ReferencedDomain = tmpReferencedDomain;
			}
        }
        else __leave;
    }

    // 
    // Indicate success.
    // 
    bSuccess=TRUE;

    } // finally
    __finally 
	{

    // 
    // Cleanup and indicate failure, if appropriate.
    // 

	if (ReferencedDomain)
		HeapFree(GetProcessHeap(), 0, ReferencedDomain);

    if(!bSuccess)
	{
        if(*Sid)
		{
            HeapFree(GetProcessHeap(), 0, *Sid);
            *Sid = NULL;
        }
	}

    } // finally

    return bSuccess;
}

```

Reply:
OK, I can now see that if the initial HeapAlloc is good and LookupAccountName fails with ERROR_INSUFFICIENT_BUFFER it will do the HeapRealloc

Reply:
I think my only problem now is at
 `	if (ReferencedDomain)`
`		HeapFree(GetProcessHeap(), 0, ReferencedDomain);`
I'm getting
`Using uninitialized memory 'ReferencedDomain'` @  `	if (ReferencedDomain)`

Reply:
I think you missed setting ReferencedDomain to NULL when you free it above:
```
			{
				HeapFree(GetProcessHeap(), 0, ReferencedDomain);
				__leave;
			}
```

If that executes, the finally block will try to free it again.  Try this:
```
			{
				HeapFree(GetProcessHeap(), 0, ReferencedDomain);
				ReferencedDomain = NULL;
				__leave;
			}
```

Reply:
That's it :)
Thanks

Reply:
It's only a problem because I have the HeapFree there. With that gone it is all OK.
As Jeff has pointed out, I can leave those to the later cleanup :)